Admin/Test-AMSI.ps1

```diff
@@ -1,144 +1,793 @@
+    [Parameter(ParameterSetName = 'RestartIIS', Mandatory = $false, ValueFromPipeline = $true)]
+    [string[]]$ServerList = $null,
+    [Parameter(ParameterSetName = 'TestAMSIAll', Mandatory = $true)]
+    [Parameter(ParameterSetName = 'CheckAMSIConfigAll', Mandatory = $true)]
+    [Parameter(ParameterSetName = 'EnableAMSIAll', Mandatory = $true)]
+    [Parameter(ParameterSetName = 'DisableAMSIAll', Mandatory = $true)]
+    [Parameter(ParameterSetName = 'RestartIISAll', Mandatory = $true)]
+    [switch]$AllServers,
+    [Parameter(ParameterSetName = 'TestAMSIAll', Mandatory = $false)]
+    [Parameter(ParameterSetName = 'CheckAMSIConfigAll', Mandatory = $false)]
+    [Parameter(ParameterSetName = 'EnableAMSIAll', Mandatory = $false)]
+    [Parameter(ParameterSetName = 'DisableAMSIAll', Mandatory = $false)]
+    [Parameter(ParameterSetName = 'RestartIISAll', Mandatory = $false)]
+    [string[]]$Sites = $null,
     [Parameter(ParameterSetName = 'RestartIIS', Mandatory = $false)]
-    [switch]$RestartIIS
+    [Parameter(ParameterSetName = 'RestartIISAll', Mandatory = $false)]
+    [switch]$Force
 )
 
-function Confirm-Administrator {
-    $currentPrincipal = New-Object Security.Principal.WindowsPrincipal( [Security.Principal.WindowsIdentity]::GetCurrent() )
-    if ($currentPrincipal.IsInRole( [Security.Principal.WindowsBuiltInRole]::Administrator )) {
-        return $true
-    } else {
-        return $false
-    }
-}
+begin {
 
-function SetCertificateValidationBehavior {
-    [CmdletBinding()]
-    param (
-        [Parameter(ParameterSetName = "Ignore", Mandatory = $true)]
-        [switch]
-        $Ignore,
-
-        [Parameter(ParameterSetName = "Default", Mandatory = $true)]
-        [switch]
-        $Default
-    )
-
-    if ($Ignore) {
-        Add-Type @"
-        using System;
-        using System.Net;
-        using System.Net.Security;
-        using System.Security.Cryptography.X509Certificates;
-        public class ServerCertificateValidationBehavior
-        {
-            public static void Ignore()
-            {
-                ServicePointManager.ServerCertificateValidationCallback +=
-                    delegate(Object obj, X509Certificate certificate, X509Chain chain, SslPolicyErrors errors)
-                    {
-                        return true;
-                    };
-            }
+    . $PSScriptRoot\..\Shared\Confirm-Administrator.ps1
+    . $PSScriptRoot\..\Shared\Confirm-ExchangeShell.ps1
+    . $PSScriptRoot\..\Shared\Invoke-ScriptBlockHandler.ps1
+    . $PSScriptRoot\..\Shared\CertificateFunctions\Enable-TrustAnyCertificateCallback.ps1
+
+    function Get-ExchangeInstallPath {
+        param(
+            [Parameter(Mandatory = $false)]
+            [string]$ExchangeServer
+        )
+        $getMSIInstallPathSB = { (Get-ItemProperty -Path Registry::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\ExchangeServer\v15\Setup -ErrorAction SilentlyContinue).MsiInstallPath }
+        if ($ExchangeServer) {
```

Comment:
This is not needed as `Invoke-ScriptblockHandler` contains the logic to determine if the call is executed against the local or remote machine. 

`Invoke-ScriptBlockHandler -ComputerName $ExchangeServer -ScriptBlock { (Get-ItemProperty -Path Registry::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\ExchangeServer\v15\Setup -ErrorAction SilentlyContinue).MsiInstallPath }`

should do the job